crypto/rsa/rsa_lib.c

```diff
@@ -163,6 +163,139 @@ void *RSA_get_ex_data(const RSA *r, int idx)
     return CRYPTO_get_ex_data(&r->ex_data, idx);
 }
 
+/*
+ * Define a scaling constant for our fixed point arithmetic.
+ * This value must be a power of two because the base two logarithm code
+ * makes this assumption.  The exponent must also be a multiple of three so
+ * that the scale factor has an exact cube root.  Finally, the scale factor
+ * should not be so large that a multiplication of two scaled numbers
+ * overflows a 64 bit unsigned integer.
+ */
+static const unsigned int scale = 1 << 18;
+static const unsigned int cbrt_scale = 1 << (2 * 18 / 3);
+
+/* Define some constants, none exceed 32 bits */
+static const unsigned int log_2  = 0x02c5c8;    /* scale * log(2) */
+static const unsigned int log_e  = 0x05c551;    /* scale * log2(M_E) */
+static const unsigned int c1_923 = 0x07b126;    /* scale * 1.923 */
+static const unsigned int c4_690 = 0x12c28f;    /* scale * 4.690 */
+
+/*
+ * Multiply two scale integers together and rescale the result.
+ */
+static ossl_inline uint64_t mul2(uint64_t a, uint64_t b)
+{
+    /*
+     * If correct rounded is desired, enable this define.
+     * Correct rounding being:
+     *      1) if the fractional part is less than one half, round down.
+     *      2) if the fractional part is greater than one half, round up.
+     *      3) if the fractional part is exactly one half, round to even.
+     * This very slightly increases accuracy.
+     */
+#ifdef OSSL_CORRECT_ROUNDING
+    uint64_t m = a * b;
+    uint32_t f = m & (scale - 1);
+
+    m /= scale;
+    if (f == scale / 2)         /* Round to even */
+        m &= ~1;
+    else if (f >= scale / 2)    /* Round up */
+        m++;
+    return m;
+#else
+    return a * b / scale;
+#endif
+}
+
+/*
+ * Calculate the cube root of a 64 bit scaled integer.
+ * Although the cube root of a 64 bit number does fit into a 32 bit unsigned
+ * integer, this is not guaranteed after scaling, so this function has a
+ * 64 bit return.  This uses the shifting nth root algorithm.
+ */
+static uint64_t icbrt64(uint64_t x)
+{
+    uint32_t r = 0;
+    uint64_t b;
+    int s;
+
+    for (s = 63; s >= 0; s -= 3) {
+        r += r;
+        b = 3 * r * (uint64_t)(r + 1) + 1;
+        if ((x >> s) >= b) {
+            x -= b << s;
+            r++;
+        }
+    }
```

Comment:
I tried matching this up against:
https://en.wikipedia.org/wiki/Shifting_nth_root_algorithm

And failed miserably. Can you help me out?

Reply:
I'll try.  Because I'm working in binary, B=2.  Because it is a cube root, n=3.  Go down to the [summary](https://en.wikipedia.org/wiki/Shifting_nth_root_algorithm#Summary) in the Wikipedia page.

I'm updating the number, x, by extracting the cube of the bits we've figured out.  This means that α is then x >> s rather than (x >> s) & 7 which simplifies things a little.  Since we're binary, β is either 0 or 1.  We need to figure out the threshold and either do nothing (β=0) or subtract the threshold out (β=1).  We also need to increment r if β=1 and shift r one bit each iteration.

Step 2 of the summary has:
 (By + β)<sup>n</sup> - B<sup>n</sup>y<sup>n</sup>
= (2y + β)<sup>3</sup> - 2<sup>3</sup>y<sup>3</sup>

Assume β=1, this becomes:
 (2y + 1)<sup>3</sup> - 8y<sup>3</sup>
= 8y<sup>3</sup> + 12<sup>2</sup> + 6y + 1 - 8y<sup>3</sup>
= 12y<sup>2</sup> + 6y + 1

Look at the expression for b in the function: b = 3r(r+1)+1

However r has just been doubled so b is effectively:
6r(2r+1)+1
= 12r<sup>2</sup> + 6r + 1

Doubling r before calculating b is an optimisation.

Anyway, the rest should fall out :)

Reply:
Ah! Thanks. And suddenly it all makes sense.